hazelcast/src/main/java/com/hazelcast/map/impl/query/PartitionScanRunner.java

```diff
@@ -94,20 +97,24 @@ public void run(String mapName, Predicate predicate, int partitionId, Result res
 
             @Override
             public void accept(Data key, Record record) {
-                Metadata metadata = PartitionScanRunner.this.getMetadataFromRecord(recordStore, key, record);
-                Object value = PartitionScanRunner.this.toData(useCachedValues
-                        ? Records.getValueOrCachedValue(record, serializationService)
-                        : record.getValue());
+                Object value = useCachedValues ? getValueOrCachedValue(record, ss) : record.getValue();
+                // TODO how can a value be null?
                 if (value == null) {
                     return;
                 }
 
-                queryEntry.init(serializationService, key, value, extractors);
-                queryEntry.setMetadata(metadata);
+                // always copy key&value to heap
+                key = toHeapData(key);
+                value = nativeMemory ? toHeapData((Data) value) : value;
+
+                queryEntry.init(ss, key, value, extractors);
                 queryEntry.setRecord(record);
                 queryEntry.setStoreAdapter(storeAdapter);
-                boolean valid = predicate.apply(queryEntry);
-                if (valid && compareAnchor(pagingPredicate, queryEntry, nearestAnchorEntry)) {
+                queryEntry.setMetadata(PartitionScanRunner.this.getMetadataFromRecord(recordStore, key, record));
+
+                if (predicate.apply(queryEntry)
```

Comment:
In order to avoid the crash, we copy key/value to heap eagerly. However, as the predicate is applied here, it seems that copying could be avoided for the discarded entries. That is, first apply the predicate to offheap data, and copy it to heap only if the predicate evaluates to `true`.
@ahmetmircik @taburet What do you think?

Reply:
makes sense to me, updated PR.